tests/e2e/pkg/util/zip.go

```diff
@@ -0,0 +1,76 @@
+/*
+Copyright 2021 The Kubernetes Authors.
+
+Licensed under the Apache License, Version 2.0 (the "License");
+you may not use this file except in compliance with the License.
+You may obtain a copy of the License at
+
+    http://www.apache.org/licenses/LICENSE-2.0
+
+Unless required by applicable law or agreed to in writing, software
+distributed under the License is distributed on an "AS IS" BASIS,
+WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+See the License for the specific language governing permissions and
+limitations under the License.
+*/
+
+package util
+
+import (
+	"archive/zip"
+	"bytes"
+	"fmt"
+	"io"
+	"io/ioutil"
+	"os"
+	"path/filepath"
+	"strings"
+)
+
+// UnzipToTempDir will decompress the provided bytes into a temporary directory that is returned
+func UnzipToTempDir(data []byte) (string, error) {
+	reader, err := zip.NewReader(bytes.NewReader(data), int64(len(data)))
+	if err != nil {
+		return "", err
+	}
+	dir, err := ioutil.TempDir("", "")
+	if err != nil {
+		return "", err
+	}
+
+	for _, r := range reader.File {
+		fileNamePath := filepath.Join(dir, r.Name)
+		if !strings.HasPrefix(fileNamePath, filepath.Clean(dir)+string(os.PathSeparator)) {
+			return "", fmt.Errorf("invalid file path: %v", fileNamePath)
+		}
+
+		if r.FileInfo().IsDir() {
+			os.MkdirAll(fileNamePath, os.ModePerm)
+			continue
+		}
+
+		if err = os.MkdirAll(filepath.Dir(fileNamePath), os.ModePerm); err != nil {
+			return "", err
+		}
+
+		output, err := os.OpenFile(fileNamePath, os.O_WRONLY|os.O_CREATE|os.O_TRUNC, r.Mode())
+		if err != nil {
+			return "", err
+		}
+
+		fileReader, err := r.Open()
+		if err != nil {
+			return "", err
+		}
+
+		_, err = io.Copy(output, fileReader)
+
+		output.Close()
+		fileReader.Close()
```

Comment:
Should we put this in a defer stack? Seems to me if `r.open` fails then `output` will not be closed and can lead to good number of open files in the system.

Reply:
If we defer these, they wont be closed until UnzipToTempDir returns which means a zip containing many files could lead to a stack overflow. Given that the kubetest2-kops process will exit if `r.open` fails, I think not deferring these is acceptable.

Reply:
Ah I didn't have enough context on kubetest2-kops process. The stack overflow part makes sense too. Thanks for clearing it up :D